Client/Frontend/Browser/Authenticator.swift

```diff
@@ -48,6 +73,25 @@ class Authenticator {
         return self.promptForUsernamePassword(viewController, credentials: nil, protectionSpace: challenge.protectionSpace, loginsHelper: nil)
     }
 
+    private static func findValidLoginAmongstLogins(logins: Cursor<LoginData>, forProtectionSpace protectionSpace: NSURLProtectionSpace) -> LoginData? {
+        // A valid login is defined as one that matches the challenge protection space and is not malformed
+        return logins.filter { login in
+            return (login?.protectionSpace.`protocol` == protectionSpace.`protocol`) && !(login?.hasMalformedHostname ?? false)
+        } .flatMap { $0 } .first
```

Comment:
Add to `ArrayExtensions`'s extension on `SequenceType`:

```
public func find(f: (Self.Generator.Element) -> Bool) -> Self.Generator.Element? {
    for x in self {
        if f(x) {
            return x
        }
    }
    return nil
}
```

At this point you can also kill the free function `find([T], f: T->Bool) -> T?` that's hiding in `Functions.swift`.

Then here just do

```
return logins.find { login in … }
```